src/libstrongswan/plugins/openssl/openssl_kem.c

```diff
@@ -0,0 +1,360 @@
+/*
+ * Copyright (C) 2024 Amazon.com Inc. or its affiliates.
+ *
+ * Copyright (C) secunet Security Networks AG
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2 of the License, or (at your
+ * option) any later version.  See <http://www.fsf.org/copyleft/gpl.txt>.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * for more details.
+ */
+
+
+#include "openssl_kem.h"
+
+#include <openssl/evp.h>
+
+#ifdef OPENSSL_IS_AWSLC
+#include <openssl/experimental/kem_deterministic_api.h>
+
+#include <crypto/key_exchange.h>
+#include <crypto/drbgs/drbg.h>
+#include <threading/thread_value.h>
+#include <utils/chunk.h>
+#include <utils/utils/object.h>
+
+typedef struct private_openssl_kem_t private_openssl_kem_t;
+
+/**
+ * Private data of an openssl_kem_t object.
+ */
+struct private_openssl_kem_t {
+	/**
+	 * Public openssl_kem_t interface.
+	 */
+	openssl_kem_t public;
+
+	/**
+	 * Key Exchange Method Transform ID.
+	 */
+	key_exchange_method_t group;
+
+	/**
+	 * OpenSSL EVP_PKEY object for a KEM Keypair. Only set on initiator.
+	 */
+	EVP_PKEY *pkey;
+
+	/**
+	 * Shared secret
+	 */
+	chunk_t shared_secret;
+
+	/**
+	 * Ciphertext. Used as the "public key" for the responder.
+	 */
+	chunk_t ciphertext;
+
+	/**
+	 * Thread-specific Deterministic Random Bit Generator (DRBG)
+	 */
+	thread_value_t *thread_drbg;
+};
+
+static int openssl_kem_get_nid(private_openssl_kem_t *this)
+{
+	switch (this->group) {
+		case ML_KEM_512:
+			return NID_MLKEM512;
+			break;
+		case ML_KEM_768:
+			return NID_MLKEM768;
+			break;
+		case ML_KEM_1024:
+			return NID_MLKEM1024;
+			break;
+		default:
+			return NID_undef;
+	}
+}
+
+METHOD(key_exchange_t, get_method, key_exchange_method_t,
+			 private_openssl_kem_t *this)
+{
+	return this->group;
+}
+
+static bool openssl_kem_generate_pkey(private_openssl_kem_t *this)
+{
+	EVP_PKEY_CTX *ctx = NULL;
+	size_t seed_length = 0;
+	chunk_t seed = chunk_empty;
+	drbg_t *drbg = this->thread_drbg->get(this->thread_drbg);
+
+	ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_KEM, NULL);
+	if (!ctx)
+	{
+		return FALSE;
+	}
+	if (!EVP_PKEY_CTX_kem_set_params(ctx, openssl_kem_get_nid(this))) {
+		EVP_PKEY_CTX_free(ctx);
+		return FALSE;
+	}
+	if (!EVP_PKEY_keygen_init(ctx))
+	{
+		EVP_PKEY_CTX_free(ctx);
+		return FALSE;
+	}
+	if (drbg)
+	{
+		if (!EVP_PKEY_keygen_deterministic(ctx, &this->pkey, NULL, &seed_length))
+		{
+			EVP_PKEY_CTX_free(ctx);
+			EVP_PKEY_free(this->pkey);
```

Comment:
Does this function actually modify the key for this use case (determining the seed length) at all (and especially if it fails)?

More importantly, the key is freed in `destroy()` anyway, which probably results in a double free if any of these function calls fails (the pointer is not set to NULL). So please remove all these `EVP_PKEY_free(this->pkey)` calls.